qualtran/bloqs/block_encoding/tensor_product.py

```diff
@@ -0,0 +1,154 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from functools import cached_property, reduce
+from typing import Dict, Sequence, Tuple
+
+from attrs import evolve, field, frozen, validators
+
+from qualtran import (
+    bloq_example,
+    BloqBuilder,
+    BloqDocSpec,
+    QAny,
+    QDType,
+    Register,
+    Signature,
+    SoquetT,
+)
+from qualtran.bloqs.block_encoding import BlockEncoding
+from qualtran.bloqs.block_encoding.lcu_select_and_prepare import PrepareOracle
+from qualtran.bloqs.bookkeeping import Partition
+from qualtran.symbolics import SymbolicFloat
+
+
+@frozen
+class TensorProduct(BlockEncoding):
+    r"""Tensor product of a sequence of block encodings.
+
+    Builds the block encoding as
+    $$
+        B[U_1 ⊗ U_2 ⊗ \cdots ⊗ U_n] = B[U_1] ⊗ B[U_2] ⊗ \cdots ⊗ B[U_n]
+    $$
+
+    When each $B[U_i]$ is a $(\alpha_i, a_i, \epsilon_i)$-block encoding of $U_i$, we have that
+    $B[U_1 ⊗ \cdots ⊗ U_n]$ is a $(\prod_i \alpha_i, \sum_i a_i, \sum_i \alpha_i \epsilon_i)$-block
+    encoding of $U_1 ⊗ \cdots ⊗ U_n$.
+
+    Args:
+        U: A sequence of block encodings.
+
+    Registers:
+        system: The system register.
+        ancilla: The ancilla register.
+        resource: The resource register.
+
+    References:
+        [Quantum algorithms: A survey of applications and end-to-end complexities](https://arxiv.org/abs/2310.03011). Dalzell et al. (2023). Ch. 10.2.
+    """
+
+    U: Sequence[BlockEncoding] = field(
```

Comment:
Does it make sense to change `U` to be a Union type of `HasLength` and `Tuple[BlockEncoding, ...]` so we can support a symbolic number of block encodings?

Reply:
I'm not really sure how that would work. We can't compute parameters such as the `alpha` of the tensor product without concretely knowing its components. Just knowing the number of components seems to not be enough.